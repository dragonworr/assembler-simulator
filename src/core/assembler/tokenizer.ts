import { InvalidTokenError } from './exceptions'

export enum TokenType {
  Comma = 'COMMA',
  String = 'STRING',
  Address = 'ADDRESS',
  Digits = 'DIGITS',
  Unknown = 'UNKNOWN'
}

export class Token {
  public type: TokenType
  public value: string
  public position: number
  public length: number

  constructor(type: TokenType, value: string, position: number) {
    this.type = type
    this.value = value
    this.position = position
    this.length = value.length
    if (type === TokenType.String || type === TokenType.Address) {
      this.length += 2
    }
  }

  public static getOriginalValue(token: Token): string {
    switch (token.type) {
      case TokenType.String:
        return `"${token.value}"`
      case TokenType.Address:
        return `[${token.value}]`
      default:
        return token.value
    }
  }
}

type Tokenizer = (input: string, current: number) => [consumedChars: number, token: Token | null]

const WHITESPACE = /\s/
const RETURN = /[\r\n]/
const TOKEN_SEPARATOR = /[\s,]/
const DIGITS = /[0-9]/
const UNKNOWN = /["0-9:a-zA-Z[\]_]/

const skipWhitespace: Tokenizer = (input, current) =>
  WHITESPACE.test(input[current]) ? [1, null] : [0, null]

const skipComment: Tokenizer = (input, current) => {
  let consumedChars = 0
  let char = input[current]
  if (char === ';') {
    consumedChars++
    char = input[current + consumedChars]
    while (char !== undefined && !RETURN.test(char)) {
      consumedChars++
      char = input[current + consumedChars]
    }
  }
  return [consumedChars, null]
}

const tokenizeCharacter =
  (type: TokenType, value: string): Tokenizer =>
  (input, current) =>
    value === input[current] ? [1, new Token(type, value, current)] : [0, null]

const tokenizeComma = tokenizeCharacter(TokenType.Comma, ',')

const tokenizeEnclosed =
  (type: TokenType, start: string, end: string): Tokenizer =>
  (input, current) => {
    let consumedChars = 0
    let char = input[current]
    if (char === start) {
      let value = ''
      consumedChars++
      char = input[current + consumedChars]
      while (char !== end) {
        if (char === undefined || RETURN.test(char)) {
          return [0, null]
        }
        value += char
        consumedChars++
        char = input[current + consumedChars]
      }
      consumedChars++
      return [consumedChars, new Token(type, value, current)]
    }
    return [0, null]
  }

const tokenizeString = tokenizeEnclosed(TokenType.String, '"', '"')
const tokenizeAddress = tokenizeEnclosed(TokenType.Address, '[', ']')

const tokenizePattern =
  (type: TokenType, pattern: RegExp): Tokenizer =>
  (input, current) => {
    let consumedChars = 0
    let char = input[current]
    if (pattern.test(char)) {
      let value = ''
      do {
        value += char
        consumedChars++
        char = input[current + consumedChars]
      } while (char !== undefined && pattern.test(char))
      if (char === undefined || TOKEN_SEPARATOR.test(char)) {
        return [consumedChars, new Token(type, value, current)]
      }
    }
    return [0, null]
  }

const tokenizeDigits = tokenizePattern(TokenType.Digits, DIGITS)
const tokenizeUnknown = tokenizePattern(TokenType.Unknown, UNKNOWN)

const tokenizers = [
  skipWhitespace,
  skipComment,
  tokenizeComma,
  tokenizeString,
  tokenizeAddress,
  tokenizeDigits,
  tokenizeUnknown
]

export const tokenize = (input: string): Token[] => {
  let currentIndex = 0
  const tokens: Token[] = []
  while (currentIndex < input.length) {
    const isTokenized = tokenizers.some(tokenizer => {
      const [consumedChars, token] = tokenizer(input, currentIndex)
      currentIndex += consumedChars
      if (token !== null) {
        if (token.type === TokenType.Address || token.type === TokenType.Unknown) {
          token.value = token.value.toUpperCase()
        }
        tokens.push(token)
      }
      return consumedChars > 0
    })
    if (!isTokenized) {
      throw new InvalidTokenError(input[currentIndex], currentIndex)
    }
  }
  return tokens
}
