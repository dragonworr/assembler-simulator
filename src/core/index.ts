export { assemble } from './assembler'
export * from './memory'
export * from './cpu'
