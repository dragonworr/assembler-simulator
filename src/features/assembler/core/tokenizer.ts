import type { SourceRange } from './types'
import { Mnemonic } from '@/common/constants'
import { trimBrackets, call } from '@/common/utils'

export enum TokenType {
  Whitespace = 'Whitespace',
  Comment = 'Comment',
  Colon = 'Colon',
  Comma = 'Comma',
  Digits = 'Digits',
  Register = 'Register',
  Address = 'Address',
  String = 'String',
  Unknown = 'Unknown'
}

export interface Token {
  type: TokenType
  value: string
  raw: string
  range: SourceRange
}

const createToken = (type: TokenType, value: string, from: number): Token => {
  const tokenValue = call((): string => {
    switch (type) {
      case TokenType.Register:
      case TokenType.Unknown:
        return value.toUpperCase()
      case TokenType.Address:
        return trimBrackets(value).trim().toUpperCase()
      case TokenType.String:
        // escape quotes and backslashes
        return JSON.parse(value)
      default:
        return value
    }
  })
  const to = from + value.length
  return {
    type,
    value: tokenValue,
    raw: value,
    range: { from, to }
  }
}

interface Tokenizer {
  reset: (source: string) => void
  next: () => Token | null
}

const createTokenizer = (rules: Record<TokenType, RegExp>): Tokenizer => {
  const ruleEntries = Object.entries(rules)
  const regexp = new RegExp(ruleEntries.map(([, pattern]) => `(${pattern.source})`).join('|'), 'y')
  let buffer = ''
  return {
    reset: source => {
      buffer = source
      regexp.lastIndex = 0
    },
    next: () => {
      const startIndex = regexp.lastIndex
      const match = regexp.exec(buffer)
      if (match !== null && match.index === startIndex) {
        for (let ruleIndex = 0; ; ruleIndex++) {
          if (match[ruleIndex + 1] !== undefined) {
            const [type] = ruleEntries[ruleIndex]
            return createToken(type as TokenType, match[0], match.index)
          }
        }
      }
      regexp.lastIndex = startIndex
      // istanbul ignore next
      if (startIndex < buffer.length) {
        throw new Error(`Unexpected token '${buffer[startIndex]}' at index ${startIndex}.`)
      }
      return null
    }
  }
}

/* eslint-disable prettier/prettier */

const tokenizer = createTokenizer({
  [TokenType.Whitespace]: /\s+/,
  [TokenType.Comment]:    /;.*/,
  [TokenType.Colon]:      /:/,
  [TokenType.Comma]:      /,/,
  [TokenType.Digits]:     /\d+\b/,
  [TokenType.Register]:   /[a-dA-D][lL]\b/,
  [TokenType.Address]:    /\[.*?\]/,
  [TokenType.String]:     /"(?:[^\\\r\n]|\\.)*?"/,
  [TokenType.Unknown]:    /[^\s;:,["]+|\[.*?(?=\s*?(?:[\r\n;:,]|$))|".*/
})

/* eslint-enable prettier/prettier */

export const tokenize = (input: string): Token[] => {
  tokenizer.reset(input)
  const tokens: Token[] = []
  let token: Token | null
  while ((token = tokenizer.next()) !== null) {
    if (token.type !== TokenType.Whitespace && token.type !== TokenType.Comment) {
      tokens.push(token)
      if (token.value === Mnemonic.END) {
        break
      }
    }
  }
  return tokens
}
