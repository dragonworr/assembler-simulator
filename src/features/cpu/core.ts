import { produce } from 'immer'
import {
  InvalidRegisterError,
  RunBeyondEndOfMemory,
  // StackOverflowError,
  // StackUnderflowError,
  DivideByZeroError
} from '../../common/exceptions'
import { Opcode, Register } from '../../common/constants'
import { Head, unsign8bit, exp } from '../../common/utils'

type GPR = [AL: number, BL: number, CL: number, DL: number]

const MAX_SP = 0xbf

enum Flag {
  Zero,
  Overflow,
  Sign,
  Interrupt
}

type SR = [zero: boolean, overflow: boolean, sign: boolean, interrupt: boolean]

interface CPU {
  gpr: GPR
  ip: number
  sp: number
  sr: SR
  isHalted: boolean
}

export const init = (): CPU => {
  return {
    gpr: [0, 0, 0, 0],
    ip: 0,
    sp: MAX_SP,
    sr: [false, false, false, false],
    isHalted: false
  }
}

const checkGPR = (register: number): Register => {
  if (register < 0 || register > 3) {
    throw new InvalidRegisterError(register)
  }
  return register
}

const checkIP = (address: number): number => {
  if (address > 0xff) {
    throw new RunBeyondEndOfMemory()
  }
  return address
}

// const checkSP = (address: number): number => {
//   if (address < 0) {
//     throw new StackOverflowError()
//   }
//   if (address > MAX_SP) {
//     throw new StackUnderflowError()
//   }
//   return address
// }

// const getFlagsValue = (sr: SR): number =>
//   sr.reduce((result, isSet, flag) => (isSet ? result + 0b10 ** flag : result), 0)

const checkOperationResult = (
  result: number,
  previousValue: number
): [result: number, flags: Partial<SR>] => {
  const finalResult = result > 0xff ? result % 0x100 : unsign8bit(result)
  const flags: Head<SR> = [false, false, false]
  if (finalResult === 0) {
    flags[Flag.Zero] = true
  } else if (finalResult >= 0x80) {
    flags[Flag.Sign] = true
  }
  if (
    /* addition */ (previousValue < 0x80 && result >= 0x80) ||
    /* subtraction */ (previousValue >= 0x80 && result < 0x80)
  ) {
    flags[Flag.Overflow] = true
  }
  return [finalResult, flags]
}

export const step = (__cpu: CPU, __memory: number[]): [cpu: CPU, memory: number[]] =>
  produce([__cpu, __memory], ([cpu, memory]) => {
    const getGPR = (register: Register): number => cpu.gpr[register]
    const setGPR = (register: Register, value: number): void => {
      cpu.gpr[register] = value
    }

    const setIP = (address: number): void => {
      cpu.ip = address
    }
    const incIP = (): void => {
      setIP(checkIP(cpu.ip + 1))
    }

    const setSR = (flags: Partial<SR>): void => {
      Object.assign(cpu.sr, flags)
    }

    const loadFromMemory = (address: number): number => {
      return memory[address]
    }
    // const storeToMemory = (address: number, machineCode: number): void => {
    //   memory[address] = machineCode
    // }

    /**
     * @modifies {@link cpu.ip}
     */
    const getNextOpcode = (): number => {
      incIP()
      return loadFromMemory(cpu.ip)
    }

    const opcode = loadFromMemory(cpu.ip)

    switch (opcode) {
      case Opcode.END:
        cpu.isHalted = true
        break
      case Opcode.ADD_REG_TO_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) + getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.SUB_REG_FROM_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) - getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.MUL_REG_BY_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) * getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.DIV_REG_BY_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          exp<number>(() => {
            const divisor = getGPR(srcReg)
            if (divisor === 0) {
              throw new DivideByZeroError()
            }
            return Math.floor(getGPR(destReg) / divisor)
          }),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.INC_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(getGPR(destReg) + 1, getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.DEC_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(getGPR(destReg) - 1, getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.MOD_REG_BY_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          exp<number>(() => {
            const divisor = getGPR(srcReg)
            if (divisor === 0) {
              throw new DivideByZeroError()
            }
            return getGPR(destReg) % divisor
          }),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.AND_REG_WITH_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) & getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.OR_REG_WITH_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) | getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.XOR_REG_WITH_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) ^ getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.NOT_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(~getGPR(destReg), getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.ROL_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          exp<number>(() => {
            const value = getGPR(destReg)
            const MSB = Math.floor(value / 0x80)
            return (value << 1) + MSB
          }),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.ROR_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          exp<number>(() => {
            const value = getGPR(destReg)
            const LSB = value % 2
            return LSB * 0x80 + (value >> 1)
          }),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.SHL_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(getGPR(destReg) << 1, getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.SHR_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(getGPR(destReg) >> 1, getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
    }
  })
