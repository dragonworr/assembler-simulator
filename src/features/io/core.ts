export const SKIP = -1

export const MAX_PORT = 0x0f

export enum InputPort {
  SimulatedKeyboard = 0,
  Thermostat = 3,
  Keyboard = 7,
  NumericKeypad = 8
}

export type InputData =
  | {
      content: number
      port: InputPort
    }
  | {
      content: null
      port: null
    }

export interface InputSignals {
  data: InputData
  interrupt: boolean
  // TODO: interruptVectorAddress
}

export const initialInputSignals: InputSignals = {
  data: {
    content: null,
    port: null
  },
  interrupt: false
}

export enum OutputPort {
  TrafficLights = 1,
  SevenSegmentDisplay = 2,
  Heater = 3,
  SnakeInMaze = 4,
  StepperMotor = 5,
  Lift = 6,
  Keyboard = 7,
  NumericKeypad = 8
}

interface OutputData {
  content: number
  port: OutputPort
}

export type OutputSignals = Partial<{
  halted: true
  requiredInputPort: InputPort
  data: OutputData
  closeWindows: true
}>

export interface Signals {
  input: InputSignals
  output: OutputSignals
}
