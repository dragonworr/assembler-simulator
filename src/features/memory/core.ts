import type { AddressToMachineCodeMap, AddressToStatementMap } from '../assembler/core'
import { Mnemonic } from '../../common/constants'

export type MemoryData = number[]

export const VDU_START_ADDRESS = 0xc0

export const initData = (): MemoryData =>
  Array.from({ length: 0x100 }, (_, address) => (address < VDU_START_ADDRESS ? 0 : 0x20))

export const initDataFrom = (map: AddressToMachineCodeMap): MemoryData => {
  const data = initData()
  for (const address in map) {
    data[address] = map[address]
  }
  return data
}

export const getSourceFrom = (map: AddressToStatementMap): string[] => {
  const source: string[] = Array.from({ length: 0x100 }, (_, address) =>
    address < VDU_START_ADDRESS ? Mnemonic.END : ''
  )
  for (const address in map) {
    const addressNumber = Number.parseInt(address)
    const statement = map[address]
    const { instruction, operands } = statement
    if (instruction.mnemonic === Mnemonic.DB) {
      const operand = operands[0]
      // if (operand.type === OperandType.Number)
      if (typeof operand.value === 'number') {
        source[address] = operand.rawValue
      } else {
        // if (operand.type === OperandType.String)
        operand.rawValue.split('').forEach((char, index) => {
          source[addressNumber + index] = char
        })
      }
    } else {
      source[address] = instruction.mnemonic
      const nextAddress = addressNumber + 1
      operands.forEach((operand, index) => {
        const { rawValue } = operand
        // Address or RegisterAddress
        source[nextAddress + index] = operand.type.endsWith('Address') ? `[${rawValue}]` : rawValue
      })
    }
  }
  return source
}
