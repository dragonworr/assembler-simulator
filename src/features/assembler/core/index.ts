import { produce } from 'immer'
import { tokenize } from './tokenizer'
import { OperandType, Statement, parse } from './parser'
import {
  DuplicateLabelError,
  AssembleEndOfMemoryError,
  LabelNotExistError,
  JumpDistanceError
} from '../../../common/exceptions'
import { Mnemonic } from '../../../common/constants'
import { call } from '../../../common/utils'

export * from './types'
export * from './tokenizer'
export * from './parser'

type LabelToAddressMap = Record<string, number>

const getLabelToAddressMap = (statements: Statement[]): LabelToAddressMap => {
  const [, labelToAddressMap] = statements.reduce<[number, LabelToAddressMap]>(
    ([address, labelToAddressMap], statement, index) => {
      const { label, instruction, operands, machineCodes } = statement
      const firstOperand = operands[0]
      return [
        instruction.mnemonic === Mnemonic.ORG
          ? (firstOperand.value as number)
          : call((): number => {
              const nextAddress =
                address +
                machineCodes.length +
                (firstOperand !== undefined && firstOperand.type === OperandType.Label ? 1 : 0)
              if (nextAddress > 0xff && index !== statements.length - 1) {
                throw new AssembleEndOfMemoryError(statement)
              }
              return nextAddress
            }),
        label === null
          ? labelToAddressMap
          : produce(labelToAddressMap, draft => {
              if (draft[label.identifier] !== undefined) {
                throw new DuplicateLabelError(label)
              }
              draft[label.identifier] = address
            })
      ]
    },
    [0, {}]
  )
  return labelToAddressMap
}

export type AddressToMachineCodeMap = Record<number, number>
export type AddressToStatementMap = Record<number, Statement>

type AssembleResult = [AddressToMachineCodeMap, AddressToStatementMap]

export const assemble = (input: string): AssembleResult => {
  const statements = parse(tokenize(input))
  const labelToAddressMap = getLabelToAddressMap(statements)
  const [, addressToMachineCodeMap, addressToStatementMap] = statements.reduce<
    [number, ...AssembleResult]
  >(
    ([address, addressToMachineCodeMap, addressToStatementMap], statement) => {
      const { instruction, operands, machineCodes } = statement
      const firstOperand = operands[0]
      if (instruction.mnemonic === Mnemonic.ORG) {
        return [firstOperand.value as number, addressToMachineCodeMap, addressToStatementMap]
      }
      if (firstOperand !== undefined && firstOperand.type === OperandType.Label) {
        const labelAddress = labelToAddressMap[firstOperand.rawValue]
        if (labelAddress === undefined) {
          throw new LabelNotExistError(firstOperand)
        }
        const distance = labelAddress - address
        if (distance < -128 || distance > 127) {
          throw new JumpDistanceError(firstOperand)
        }
        const unsignedDistance = distance < 0 ? 0x100 + distance : distance
        machineCodes.push(unsignedDistance)
      }
      return [
        address + machineCodes.length,
        produce(addressToMachineCodeMap, draft => {
          machineCodes.forEach((machineCode, index) => {
            draft[address + index] = machineCode
          })
        }),
        produce(addressToStatementMap, draft => {
          draft[address] = statement
        })
      ]
    },
    [0, {}, {}]
  )
  return [addressToMachineCodeMap, addressToStatementMap]
}
