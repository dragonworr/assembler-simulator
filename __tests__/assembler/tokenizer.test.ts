import { tokenize } from '../../src/features/assembler/core/tokenizer'

describe('tokenizer', () => {
  it('should skip whitespace', () => {
    expect(tokenize(' \t\ndone: \t\nend')).toMatchSnapshot()
  })

  it('should skip line with comment', () => {
    expect(tokenize('; this is a comment\nend')).toMatchSnapshot()
  })

  it('should skip comment at the end of the line', () => {
    expect(tokenize('done:; this is a comment\nend; this is another comment')).toMatchSnapshot()
  })

  it('should tokenize comma', () => {
    expect(tokenize(',\n,,')).toMatchSnapshot()
  })

  it('should tokenize digits', () => {
    expect(tokenize('0 01 002')).toMatchSnapshot()
  })

  it('should tokenize register', () => {
    expect(tokenize('al Bl cL DL')).toMatchSnapshot()
  })

  it('should tokenize address', () => {
    expect(tokenize('[] [00] [al]')).toMatchSnapshot()
  })

  it('should tokenize string', () => {
    expect(tokenize('""\n"this is a string"')).toMatchSnapshot()
  })
})
