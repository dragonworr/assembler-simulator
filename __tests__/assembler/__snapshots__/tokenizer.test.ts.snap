// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenizer should skip comment at the end of the line 1`] = `
Array [
  Object {
    "range": Object {
      "from": 0,
      "to": 4,
    },
    "raw": "done",
    "type": "Unknown",
    "value": "DONE",
  },
  Object {
    "range": Object {
      "from": 4,
      "to": 5,
    },
    "raw": ":",
    "type": "Colon",
    "value": ":",
  },
  Object {
    "range": Object {
      "from": 25,
      "to": 28,
    },
    "raw": "end",
    "type": "Unknown",
    "value": "END",
  },
]
`;

exports[`tokenizer should skip line with comment 1`] = `
Array [
  Object {
    "range": Object {
      "from": 20,
      "to": 23,
    },
    "raw": "end",
    "type": "Unknown",
    "value": "END",
  },
]
`;

exports[`tokenizer should skip whitespace 1`] = `
Array [
  Object {
    "range": Object {
      "from": 3,
      "to": 7,
    },
    "raw": "done",
    "type": "Unknown",
    "value": "DONE",
  },
  Object {
    "range": Object {
      "from": 7,
      "to": 8,
    },
    "raw": ":",
    "type": "Colon",
    "value": ":",
  },
  Object {
    "range": Object {
      "from": 11,
      "to": 14,
    },
    "raw": "end",
    "type": "Unknown",
    "value": "END",
  },
]
`;

exports[`tokenizer should tokenize address 1`] = `
Array [
  Object {
    "range": Object {
      "from": 0,
      "to": 2,
    },
    "raw": "[]",
    "type": "Address",
    "value": "",
  },
  Object {
    "range": Object {
      "from": 3,
      "to": 7,
    },
    "raw": "[00]",
    "type": "Address",
    "value": "00",
  },
  Object {
    "range": Object {
      "from": 8,
      "to": 12,
    },
    "raw": "[al]",
    "type": "Address",
    "value": "AL",
  },
]
`;

exports[`tokenizer should tokenize comma 1`] = `
Array [
  Object {
    "range": Object {
      "from": 0,
      "to": 1,
    },
    "raw": ",",
    "type": "Comma",
    "value": ",",
  },
  Object {
    "range": Object {
      "from": 2,
      "to": 3,
    },
    "raw": ",",
    "type": "Comma",
    "value": ",",
  },
  Object {
    "range": Object {
      "from": 3,
      "to": 4,
    },
    "raw": ",",
    "type": "Comma",
    "value": ",",
  },
]
`;

exports[`tokenizer should tokenize digits 1`] = `
Array [
  Object {
    "range": Object {
      "from": 0,
      "to": 1,
    },
    "raw": "0",
    "type": "Digits",
    "value": "0",
  },
  Object {
    "range": Object {
      "from": 2,
      "to": 4,
    },
    "raw": "01",
    "type": "Digits",
    "value": "01",
  },
  Object {
    "range": Object {
      "from": 5,
      "to": 8,
    },
    "raw": "002",
    "type": "Digits",
    "value": "002",
  },
]
`;

exports[`tokenizer should tokenize register 1`] = `
Array [
  Object {
    "range": Object {
      "from": 0,
      "to": 2,
    },
    "raw": "al",
    "type": "Register",
    "value": "AL",
  },
  Object {
    "range": Object {
      "from": 3,
      "to": 5,
    },
    "raw": "Bl",
    "type": "Register",
    "value": "BL",
  },
  Object {
    "range": Object {
      "from": 6,
      "to": 8,
    },
    "raw": "cL",
    "type": "Register",
    "value": "CL",
  },
  Object {
    "range": Object {
      "from": 9,
      "to": 11,
    },
    "raw": "DL",
    "type": "Register",
    "value": "DL",
  },
]
`;

exports[`tokenizer should tokenize string 1`] = `
Array [
  Object {
    "range": Object {
      "from": 0,
      "to": 2,
    },
    "raw": "\\"\\"",
    "type": "String",
    "value": "",
  },
  Object {
    "range": Object {
      "from": 3,
      "to": 21,
    },
    "raw": "\\"this is a string\\"",
    "type": "String",
    "value": "this is a string",
  },
]
`;
